### PR TITLE
feat(auth): Add advanced security feature to initiateAuth

### PR DESCRIPTION
**Pull request by royjit**
*Issue #, if available:*

*Description of changes:*

*Check points: (check or cross out if not relevant)*

- [ ] Added new tests to cover change, if needed
- [ ] Build succeeds with all target using Swift Package Manager
- [ ] All unit tests pass
- [ ] All integration tests pass
- [ ] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)
- [ ] Documentation update for the change if required
- [ ] PR title conforms to conventional commit style
- [ ] If breaking change, documentation/changelog update with migration instructions

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/StateMachine/CodeGen/Events/CredentialStoreEvent.swift [hunk 1]
```diff
@@ -11,11 +11,13 @@ import AWSPluginsCore
 enum CredentialStoreData: Codable, Equatable {
     case amplifyCredentials(AmplifyCredentials)
     case deviceMetadata(DeviceMetadata, Username)
+    case asfDeviceId(String?, Username)
```
Comment (harsh62) [inline AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/StateMachine/CodeGen/Events/CredentialStoreEvent.swift:14]: Why is deviceId an optional?

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Support/Utils/InitiateAuthInput+Amplify.swift [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+            let asfClient = environment.cognitoUserPoolASFFactory()
+            var deviceInfo: ASFDeviceBehavior
+            var appInfo: ASFAppInfoBehavior
+#if os(iOS)
+            deviceInfo = ASFUIDeviceInfo(id: asfDeviceId)
+            appInfo = ASFAppInfo()
+
+#else
+            // TODO: Add for other platforms
+#endif
```
Comment (harsh62) [inline AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Support/Utils/InitiateAuthInput+Amplify.swift:106]: We should create a task for it...

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Actions/CredentialStore/StoreCredentials.swift [hunk 3]
```diff
@@ -36,6 +36,10 @@ struct StoreCredentials: Action {
                 try amplifyCredentialStore.saveCredential(amplifyCredentials)
             case .deviceMetadata(let deviceMetadata, let username):
                 try amplifyCredentialStore.saveDevice(deviceMetadata, for: username)
+            case .asfDeviceId(let deviceId, let username):
+                if let deviceId = deviceId {
+                    try amplifyCredentialStore.saveASFDevice(deviceId, for: username)
+                }
```
Comment (harsh62) [inline AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Actions/CredentialStore/StoreCredentials.swift:42]: Shouldn't the deviceId always exist?

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Actions/CredentialStore/LoadCredentialStore.swift [hunk 4]
```diff
@@ -34,11 +34,14 @@ struct LoadCredentialStore: Action {
             let credentialStoreData: CredentialStoreData
             switch credentialStoreType {
             case .amplifyCredentials:
-                let storedCredentials = try amplifyCredentialStore.retrieveCredential() as! AmplifyCredentials
+                let storedCredentials = try amplifyCredentialStore.retrieveCredential()
                 credentialStoreData = .amplifyCredentials(storedCredentials)
             case .deviceMetadata(let username):
-                let deviceMetadata = try amplifyCredentialStore.retrieveDevice(for: username) as! DeviceMetadata
+                let deviceMetadata = try amplifyCredentialStore.retrieveDevice(for: username)
```
Comment (harsh62) [inline AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Actions/CredentialStore/LoadCredentialStore.swift:40]: Thanks for cleaning this up!!

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/ASF/CognitoUserPoolASF+KeyChain.swift [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+import Foundation
+
+extension CognitoUserPoolASF {
+
+    static func asfDeviceID(
+        for username: String,
+        environment: AuthEnvironment) async throws -> String {
+            let credentialStoreClient = environment.credentialStoreClientFactory()
+            let data = try? await credentialStoreClient.fetchData(
+            type: .asfDeviceId(username: username))
```
[the remark this answers is not in the document]
Comment (royjit) [reply AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/ASF/CognitoUserPoolASF+KeyChain.swift]: fixed

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/StateMachine/CodeGen/Events/CredentialStoreEvent.swift (hunk 1 above)
Comment (royjit) [reply AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/StateMachine/CodeGen/Events/CredentialStoreEvent.swift]: Changed to non optional

**Review by harsh62 — APPROVED** [review]
LGTM

**Pull request merged**